source/common/event/file_event_impl.h

```diff
@@ -41,5 +45,136 @@ class FileEventImpl : public FileEvent, ImplBase {
 };
+
+// Forward declare for friend class.
+class UserSpaceFileEventFactory;
+
+// The interface of populating event watcher and obtaining the active events. The events are the
+// combination of FileReadyType values. The event listener is populated by user event registration
+// and io events passively. Also the owner of this listener query the activated events by calling
+// triggeredEvents and getAndClearEphemeralEvents.
+class EventListener {
+public:
+  virtual ~EventListener() = default;
+
+  // Get the events which are enabled and triggered.
+  virtual uint32_t triggeredEvents() PURE;
+  // Get the events which are ephemerally activated. Upon returning the ephemeral events are
+  // cleared.
+  virtual uint32_t getAndClearEphemeralEvents() PURE;
+
+  /**
+   * FileEvent::setEnabled is invoked.
+   * @param enabled_events supplied the event of setEnabled.
+   */
+  virtual void onEventEnabled(uint32_t enabled_events) PURE;
+
+  /**
+   * FileEvent::activate is invoked.
+   * @param enabled_events supplied the event of activate().
+   */
+  virtual void onEventActivated(uint32_t activated_events) PURE;
+};
+
+// Return the enabled events except EV_CLOSED. This implementation is generally good since only
+// epoll supports EV_CLOSED but the entire envoy code base supports another poller. The event owner
+// must assume EV_CLOSED is never activated. Also event owner must tolerate that OS could notify
+// events which are not actually triggered.
+// TODO(lambdai): Add support of delivering EV_CLOSED.
+class DefaultEventListener : public EventListener {
+public:
+  ~DefaultEventListener() override = default;
+
+  // Return both read and write.
+  uint32_t triggeredEvents() override { return pending_events_ & (~Event::FileReadyType::Closed); }
+
+  void onEventEnabled(uint32_t enabled_events) override { pending_events_ = enabled_events; }
+
+  void onEventActivated(uint32_t activated_events) override {
+    ephermal_events_ |= activated_events;
+  }
+
+  uint32_t getAndClearEphemeralEvents() override { return std::exchange(ephermal_events_, 0); }
+
+private:
+  // The persisted interested events and ready events.
+  uint32_t pending_events_{};
+  // The events set by activate() and will be cleared after the io callback.
+  uint32_t ephermal_events_{};
+};
+
+// A FileEvent implementation which is used to drive BufferedIoSocketHandle.
+class UserSpaceFileEventImpl : public FileEvent, Logger::Loggable<Logger::Id::io> {
+public:
+  ~UserSpaceFileEventImpl() override {
+    if (schedulable_.enabled()) {
+      schedulable_.cancel();
+    }
+  }
+
+  // Event::FileEvent
+  void activate(uint32_t events) override {
+    // Only supported event types are set.
+    ASSERT((events & (FileReadyType::Read | FileReadyType::Write | FileReadyType::Closed)) ==
+           events);
+    event_listener_.onEventActivated(events);
+    if (!schedulable_.enabled()) {
+      schedulable_.scheduleCallbackNextIteration();
```

Comment:
@antoniovicente @yanavlasov Looks like you both have the agreement that `activate() could be called from any thread`. 
Before and after PR11750 there is "event_active(&event_active...)" in `FileEventImpl::active()".

Should be fix the usage of `invoke activate() from other thread`, or should we make `activate()` thread safe?

Reply:
There is no need for activate to be thread safe.

Reply:
Thanks @antoniovicente !

I am closing this conversion for now. The conclusion is we can add check to prevent the misuse(call from other thread). This code is not guaranteed to be correct under cross thread invoke. CC @yanavlasov